net: ip: 6lo (IPHC) compression rework

alexanderwachter opened the pull request:
This is a rework of the 6lo IPHC. The compression now works inline
on the original buffer instead of allocation a new one.
Additionally DAM_11 (Destination address fully elided) has precedence
over DAM_10 (16 bit compressible) now.

Commit:
net: ip: 6lo compression rework

This is a rework of the 6lo IPHC. The compression now works inline
on the original buffer instead of allocation a new one.
Additionally DAM_11 (Destination address fully elided) has precedence
over DAM_10 (16 bit compressible) now.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

rveerama1 commented on subsys/net/ip/6lo_private.h:
```diff
@@ -81,6 +81,7 @@
 #define NET_6LO_NHC_UDP_PORT_10		0x02
 #define NET_6LO_NHC_UDP_PORT_11		0x03
 
+#define NET_6LO_NCH_UDP_CHECKSUM    0x04
```
Is it typo?  "NCH", rest of the macros has NHC.

alexanderwachter replied:
Typo.

Commit:
net: ip: 6lo (IPHC) uncompression rework

This is a rework of the 6lo IPHC uncompression.
The uncompression now tries to work in place on the original buffer
instead of allocation a new one. If there is not enough tail-room,
a new buffer is allocated and filled with the IP and maybe UDP header.
The compressed header is pulled from the original buffer and the
buffer is preserved.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
net: ip: net_pkt: Add pkt_cursor_advance before contiguous check

When net_pkt_skip skips the entire data in a net buffer, the cursor
still points to this buffer on data that is off by one.
Calling pkt_cursor_advance in net_pkt_is_contiguous fixes this and
moves the cursor to the next buffer.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
net: ip: net_pkt: additional pkt_cursor_advance after cursor operations. (af268141c08c6fa44f0b024c1f9811579c9ac449) is a workaround and should be fixed in an extra PR.
@tbursztyka Maybe you can spot the problem here.

jukkar commented on subsys/net/ip/6lo_private.h:
```diff
@@ -81,6 +81,7 @@
 #define NET_6LO_NHC_UDP_PORT_10		0x02
 #define NET_6LO_NHC_UDP_PORT_11		0x03
 
+#define NET_6LO_NHC_UDP_CHECKSUM    0x04
```
Please indent the value against other values, also the empty line before could be removed too.

alexanderwachter replied:
Done

jukkar reviewed (commented):
LGTM. Just minor nitpick and also the checkpatch issue should be fixed too.

alexanderwachter commented:
@jukkar I coulnd't find any "beautiful" solution for the checkpatch error so far. Need to think about this a little more.

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1823,6 +1823,8 @@ bool net_pkt_is_contiguous(struct net_pkt *pkt, size_t size)
 	if (pkt->cursor.buf && pkt->cursor.pos) {
```
There is some garbage at the end of commit message of "net: ip: net_pkt: Add pkt_cursor_advance before contiguous check"
Also there is redundant information in the commit subject, and could be written shorter like this "net: pkt: Add pkt_cursor_advance before contiguous check"

alexanderwachter replied:
Removed.

tbursztyka commented on subsys/net/ip/6lo.c:
```diff
@@ -1474,6 +1474,8 @@ static bool uncompress_IPHC_header(struct net_pkt *pkt,
 		}
 	}
 
+	net_pkt_cursor_init(pkt);
```
you will need to split this in 2: one on subsys/net and one on tests/net

alexanderwachter replied:
Not sure here. I need to call  net_pkt_cursor_init(pkt) because i have added a new pkt in front.

tbursztyka replied:
No my point was the patch touches the net subsystem and the test: better have one patch for each.

alexanderwachter replied:
Done.

tbursztyka commented on subsys/net/ip/6lo.c:
```diff
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (c) 2016 Intel Corporation
+ * Copyright (c) 2019 Alexander Wachter
```
review the commit title, instead of 'net: ip:'  have 'subys: net:'  or just 'net:' apply that on other commit titles.

jukkar replied:
> have 'subys: net:'

There should never be "subsys: net: " prefix as the "net: " is already the subsystem so there is no need to add duplicate information.
Depending on context, sometimes "net: ip: ..." makes sense but in this case for 6lo stuff, the prefix should probably be just "net: 6lo: ..." or similar (depending on the stuff we are changing in the commit).

tbursztyka replied:
> There should never be "subsys: net: " prefix as the "net: " is already the subsystem

I have been using that format without any complaints, moreover when other patches touches tests/net, then tests: net: was the prefix, so it differentiates which part in zephyr targeting net domain.

tbursztyka reviewed (changes requested):
first patch is quite big as it is changing the overall buffer access logic. I guess it would require an actual runtime test of a 15.4 network to see if it did not break anything.

alexanderwachter commented:
Next step is to remove the hacks in the 802.15.4 code that protects the ll addresses. They are obsolete with this PR

jukkar commented:
> require an actual runtime test of a 15.4 network to see if it did not break anything

Yes, I agree here. We would need to check this in 802.15.4 and in BT.

alexanderwachter commented:
Indeed this need to be checked. Does anyone have a setup to check?

Commit:
net: ip: 6lo: Remove unnecessary net_buf allocation and memmove

This patch tries to avoid memmoves and buffer allocations when there
is enough space in the original buffer. Headers are still contiguous
but not in the same buffer.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
tests: net: 6lo: Adapt tests to deal with headers in multiple frames

The testcases are adapted to handle more than one net buffer in
a net_pkt. After the IPHC optimization it is possible that IP header
is not in the same buffer than the UDP header.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

rlubos commented:
@alexanderwachter I did a quick test of 802.15.4 and nRF52840, seems to work fine :+1:

I'll do some more thorough test (15.4 and IPSP) once you're done with decompression.

Commit:
tests: net: 6Lo: Add test criterion HDR diff

This commit adds a test criterion that checks how many bytes
are compressed.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
Reworked the uncompression too now.

rlubos commented:
Tested the new commits with 802.15.4 and BLE IPSP. 

The 802.15.4 communication works fine (tested with `echo_client`/`echo_server` samples running on two nRF52840_pca10056 boards).

Sadly I cannot say the same about BLE IPSP, it appears to be broken (same behavior observed on current master, so it's likely not related to this PR). I've tried `echo_server` and `ipsp` samples. In both cases, they do not seem to echo back the TCP data (tried `echo_client` from net-tools and telnet).
The `echo_server`, when tested with UDP packets (`echo_client` again) works fine until a 1280 byte datagram is received - it's dropped and ICMP response is sent (I'm not sure though if such large packets are supported by IPSP). 
Just to be clear - I observe the same issues on current master, so I doubt these problems relate to the changes in this PR. Unfortunately, I have no time right now to investigate it any further. Attached wireshark logs in case someone wanted to have a look.

[ble_ipsp_issues.zip](https://github.com/zephyrproject-rtos/zephyr/files/3367450/ble_ipsp_issues.zip)

tbursztyka commented:
@alexanderwachter will this PR address also not-in-place header compression? (so user would provide a buffer of its own where compressed header would be put). This would solve an issue we have with TCP where the whole tcp packet has to be cloned prior to be sent.

alexanderwachter commented:
@tbursztyka I had a look at the TCP stack and I don't think we can solve this issue with a separate buffer for compression. The problem is that if we have a separate buffer we can't insert this buffer to the pkt because if we do so we would need to pull the original header. If we don't insert the compressed buffer we need to handle TCP connections different than others in all L2 using 6Lo.
An alternative would be to recover the original header by decompressing it after the data is sent. This could be done by checking the headroom and do an in-place decompression in the buffers headroom.

jukkar commented:
We clone the TCP packet before sending the clone packet. The original packet is kept until we have received ACK from peer. It does not really matter what the 6lo does with the cloned packet.

alexanderwachter commented:
@jukkar we try to avoid cloning the packet but only increase the reference count. The problem is that at the moment we would try to compress the packet again. My preferred solution is to check if the packet is compressed already and skip the compression if so.

alexanderwachter commented:
@tbursztyka I think we should handle the TCP issue in a follow-up PR. Can you create an issue for that and assign us both?

tbursztyka commented:
> I had a look at the TCP stack and I don't think we can solve this issue with a separate buffer for compression. The problem is that if we have a separate buffer we can't insert this buffer to the pkt because if we do so we would need to pull the original header. If we don't insert the compressed buffer we need to handle TCP connections different than others in all L2 using 6Lo.

Nope, it can be solved as I planned: 6lo compression would put the result in an external buffer and let net_pkt untouched. Have a look at how 15.4 creates frames from net_pkt: it has a dedicated frame_buffer_data buffer where to create the frame. So compressed header could be put there (for the very first frame to be sent) then it would continue on the net_pkt thanks to the cursor etc...

bt l2 would require some modification as well but in the end we could totally scrap TCP net_pkt clone stuff.

tbursztyka commented on subsys/net/ip/6lo.c:
```diff
@@ -46,6 +46,74 @@ static inline u8_t get_6co_cid(struct net_icmpv6_nd_opt_6co *opt)
+static const u8_t sa_inline_size_table[] = {16, 8, 2, 0, 0, 8, 2, 0};
+
+static const u8_t da_inline_size_table[] =
+/* The first bit is M, the second DAC
```
Put this comment before declaring the variable

alexanderwachter replied:
This comment should show the borders of the values but actually, it looks strange.
Moved on top.

tbursztyka commented on subsys/net/ip/6lo.c:
```diff
@@ -46,6 +46,74 @@ static inline u8_t get_6co_cid(struct net_icmpv6_nd_opt_6co *opt)
+/* The first bit is M, the second DAC
+ *	| M=0 DAC=0 | M=0 DAC=1 | M=1 DAC=0  | M=1 DAC=1 (DAM always 00)
+ */
+	{16, 8, 2, 0, 0, 8, 2, 0, 16, 6, 4, 1, 6};
```
Move these variable at the top of the file

alexanderwachter replied:
Done

tbursztyka reviewed (commented):
minor issue in uncompression rework
